Upgrade terser-webpack-plugin to 2.3.5

jsnajdr opened the pull request:
Upgrades `terser-webpack-plugin` to the latest version. I was careful to ensure that the lockfile diff indeed updates only this package and its dependencies, and doesn't include unrelated updates to packages like `babel-*`.

The new `terser-webpack-plugin` version has better error reporting. Our `icfy-stats` job in CircleCI often fails with errors from webpack like this one:
```
ERROR in entry-gutenboarding.046ef0dc4f72b77893c6.min.js from Terser
Error: Call retries were exceeded
    at ChildProcessWorker.initialize (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:193:21)
    at ChildProcessWorker.onExit (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:263:12)
    at ChildProcess.emit (events.js:223:5)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
```

The "Call retries were exceeded" error comes from `jest-worker`, which runs a pool of parallel child processes with `terser` that consume a queue of files to minify. If the child process exits with non-zero exit code, `jest-worker` tries to run it again one or more times.

But in the error message above, we don't see why `terser` failed at all! That's what the new `terser-webpack-plugin` attempts to fix, looking at the worker process stdout and stderr and reporting it.

Previous attempt at the upgrade had to be reverted by @blowery in #39400. Because there were errors on Edge, due to incorrect transpilation of object spread syntax? I believe this is not related to `terser-webpack-plugin` at all. The unwanted lockfile updates of `babel-*` package likely caused that.

See also: #39502 where the `icfy-stats` was made to fail when errors like this happen.

jsnajdr committed:
Upgrade terser-webpack-plugin to 2.3.5

sirreal reviewed (approved):
👍 Looks good, thanks! Let's wait for e2e then 🚢

jsnajdr commented:
The `icfy-stats` job is failing. Interesting 🙂

sirreal commented:
> The `icfy-stats` job is failing. Interesting 🙂

Rerunning the job. It seems to fail intermittently 😕

jsnajdr commented:
> Rerunning the job. It seems to fail intermittently 😕

The second run failed, too. The same way. I like permanent failures, they are easier to debug than the intermittent ones 😉

jsnajdr commented:
For the record, there are three kinds of `terser-webpack-plugin` errors I've seen. The plugin uses `jest-worker` to create a pool of parallel processes that run the actual `terser`, and feeds the list of files to minify to the pool's queue.

When a worker process fails, the pool tries to restart it several times (default is 3). Two of the three errors I've seen happen when the worker process fails (in the `onExit` event callback) and `jest-worker` tries to restart it.

1. There were too many restarts already, giving up:
```
ERROR in entry-gutenboarding.046ef0dc4f72b77893c6.min.js from Terser
Error: Call retries were exceeded
    at ChildProcessWorker.initialize (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:193:21)
    at ChildProcessWorker.onExit (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:263:12)
    at ChildProcess.emit (events.js:223:5)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
```

2. When trying to respawn the worker, we fail because of insufficient memory:
```
internal/child_process.js:394
    throw errnoException(err, 'spawn');
    ^

Error: spawn ENOMEM
    at ChildProcess.spawn (internal/child_process.js:394:11)
    at spawn (child_process.js:540:9)
    at Object.fork (child_process.js:108:10)
    at ChildProcessWorker.initialize (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:137:44)
    at ChildProcessWorker.onExit (/home/circleci/wp-calypso/node_modules/jest-worker/build/workers/ChildProcessWorker.js:263:12)
    at ChildProcess.emit (events.js:223:5)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12) {
  errno: 'ENOMEM',
  code: 'ENOMEM',
  syscall: 'spawn'
}
```

3. When sending a task to a worker process, we can't write to the pipe. The worker process most likely already died:
```
events.js:200
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at ChildProcess.target._send (internal/child_process.js:806:20)
    at ChildProcess.target.send (internal/child_process.js:676:19)
    at ChildProcessWorker.send (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/workers/ChildProcessWorker.js:330:17)
    at WorkerPool.send (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/WorkerPool.js:32:34)
    at Farm._process (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/Farm.js:129:10)
    at Farm._enqueue (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/Farm.js:152:10)
    at Farm._push (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/Farm.js:159:12)
    at /home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/Farm.js:90:14
    at new Promise (<anonymous>)
    at Farm.doWork (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/Farm.js:56:12)
    at JestWorker._callFunctionWithArgs (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/index.js:178:23)
    at TaskRunner.runTask (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/dist/TaskRunner.js:41:26)
    at enqueue (/home/circleci/wp-calypso/node_modules/terser-webpack-plugin/dist/TaskRunner.js:89:35)
    at /home/circleci/wp-calypso/node_modules/terser-webpack-plugin/dist/TaskRunner.js:113:86
Emitted 'error' event on ChildProcess instance at:
    at internal/child_process.js:810:39
    at processTicksAndRejections (internal/process/task_queues.js:76:11) {
  errno: 'EPIPE',
  code: 'EPIPE',
  syscall: 'write'
}
```

The common theme is that the worker processes are unexpectedly crashing. I suspect it's because the CircleCI box is running out of memory. Not the Node.js process and its heap, but the virtual machine and it's OS itself. Our CircleCI boxes are just 2CPU/4096MB, which is not much.

sirreal commented:
It passed!

jsnajdr commented:
I reached a conclusion about why Terser is misbehaving during the icfy-analyze job: the CircleCI box it runs in is running out of memory, and the Terser worker processes get a SIGKILL signal from the kernel.

The `terser-webpack-plugin` uses the `jest-worker` library to create a pool of parallel worker processes that do the actual minification. The library swallows the errors when a worker process unexpectedly exits, doesn’t report them, and doesn’t even handle them 100% correctly. So it took some effort to figure out.

The root cause for us is that the CircleCI boxes are only 2CPU/4096MB. The `icfy-analyze` webpack build consumes more memory than the vanilla production build, that’s why it fails.

The `terser-webpack-plugin` and especially `jest-worker` could be patched to report the errors better, but other than that, there is not any bug that would be responsible for these failures and that could be fixed.

jsnajdr committed:
Disable terser-webpack-plugin parallel mode when building stats in Circle CI

Running the minifier in the main process and only in one thread should consume less
memory overall than spawning a separate worker process.

jsnajdr commented:
> It passed!

@sirreal I think it's random -- sometimes is fails, sometimes it succeeds, depending on whether the particular job happens to run out of memory or not.

sgomes commented:
Thank you for looking into this so extensively, @jsnajdr! In order to definitively solve the issue, I take it we'd need to either increase the amount of memory in our CircleCI instances, or reduce Terser memory usage somehow?

sirreal commented:
I believe if we migrate to GitHub actions, which I'd expect to be fairly straightforward for this, we'd we’d get 7GB:

> GitHub hosts Linux and Windows runners on `Standard_DS2_v2` virtual machines in Microsoft Azure

https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#about-github-hosted-runners

https://docs.microsoft.com/en-us/azure/virtual-machines/windows/sizes-general#dsv2-series (edited) 

I think that's the most interesting avenue to explore.

jsnajdr commented:
> reduce Terser memory usage somehow?

@sgomes I tried something in https://github.com/Automattic/wp-calypso/pull/39535/commits/29f8c099398432600e01dfeb39b26c7f2f7fbaca -- disable the `parallel` mode for the `terser-webpack-plugin` and make it execute `terser` inside one Node.js process, single-threaded, instead of spawning workers. That should theoretically consume less OS memory, as it saves the overhead of extra processes.

But it didn't work: the main Node.js process was killed by the `oom-killer`, exiting with code 137, which is the expected one for `SIGKILL`. In short, the error reporting is better, the result is the same 🙂

I don't have any other ideas. @sirreal suggestion to use Github Actions seems like the best one.

Or we can increase memory for the CircleCI container -- how does one do that? I don't know.

scinos commented:
Outdated, we are in `"terser-webpack-plugin": "^5.1.1"` now
